redundant_optional_initialization removes opening bracket before property observer

### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

The redundant_optional_initialization rule removes the opening bracket before a property observer which causes a syntax issue. For example 

```swift
var foo: String? = nil {
  didSet {
    print("foo")
  }
}
```

is changed to 

```swift
var foo: String?
  didSet {
    print("foo")
  }
}
```

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint lint       

Linting Swift files in current working directory
Linting 'Foo.swift' (1/1)
/Foo.swift:3:21: warning: Redundant Optional Initialization Violation: Initializing an optional variable with nil is redundant. (redundant_optional_initialization)
Done linting! Found 1 violation, 0 serious in 1 file.
```

### Environment

 ```bash
$ swiftlint version
swiftlint version
0.44.0
```

@iressler

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -22,6 +22,10 @@
 #### Bug Fixes
 
 * Fixes a bug with the `missing_docs` rule where `excludes_inherited_types` would not be set.
+* Fix redundant_optional_initialization autocorrect broken
+  in case observer's brace exists.
+  [Naruki Chigira](https://github.com/naru-jpn)
+  [#3718](https://github.com/realm/SwiftLint/issues/3718)
 
 * Fix a false positive in the `unneeded_break_in_switch` rule when
   using `do/catch`.  

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -22,6 +22,7 @@
 #### Bug Fixes
 
 * Fixes a bug with the `missing_docs` rule where `excludes_inherited_types` would not be set.
+
 * Fix redundant_optional_initialization autocorrect broken
   in case observer's brace exists.
   [Naruki Chigira](https://github.com/naru-jpn)

--- a/Source/SwiftLintFramework/Rules/Idiomatic/RedundantOptionalInitializationRule.swift
+++ b/Source/SwiftLintFramework/Rules/Idiomatic/RedundantOptionalInitializationRule.swift
@@ -85,7 +85,27 @@ public struct RedundantOptionalInitializationRule: SubstitutionCorrectableASTRul
             Example("var myVar: Int?↓=nil\n"): Example("var myVar: Int?\n"),
             Example("var myVar: Optional<Int>↓=nil\n"): Example("var myVar: Optional<Int>\n"),
             Example("class C {\n#if true\nvar myVar: Int?↓ = nil\n#endif\n}"):
-                Example("class C {\n#if true\nvar myVar: Int?\n#endif\n}")
+                Example("class C {\n#if true\nvar myVar: Int?\n#endif\n}"),
+            Example("""
+            var myVar: Int?↓ = nil {
+                didSet { }
+            }
+            """):
+                Example("""
+                var myVar: Int? {
+                    didSet { }
+                }
+                """),
+            Example("""
+            var myVar: Int?↓=nil{
+                didSet { }
+            }
+            """):
+                Example("""
+                var myVar: Int?{
+                    didSet { }
+                }
+                """)
         ]
 
         guard SwiftVersion.current >= .fourDotOne else {
@@ -104,7 +124,7 @@ public struct RedundantOptionalInitializationRule: SubstitutionCorrectableASTRul
         return corrections
     }()
 
-    private let pattern = "\\s*=\\s*nil\\b\\s*\\{?"
+    private let pattern = "(\\s*=\\s*nil\\b)\\s*\\{?"
 
     public func validate(file: SwiftLintFile, kind: SwiftDeclarationKind,
                          dictionary: SourceKittenDictionary) -> [StyleViolation] {
@@ -126,13 +146,22 @@ public struct RedundantOptionalInitializationRule: SubstitutionCorrectableASTRul
             typeIsOptional(type),
             !dictionary.enclosedSwiftAttributes.contains(.lazy),
             dictionary.isMutableVariable(file: file),
-            let range = range(for: dictionary, file: file),
-            let match = file.match(pattern: pattern, with: [.keyword], range: range).first,
-            match.location == range.location + range.length - match.length else {
+            let range = range(for: dictionary, file: file) else {
+            return []
+        }
+
+        let regularExpression = regex(pattern)
+        guard let match = regularExpression.matches(in: file.stringView, range: range).first,
+            match.range.location == range.location + range.length - match.range.length else {
                 return []
         }
 
-        return [match]
+        let matched = file.stringView.substring(with: match.range)
+        if matched.hasSuffix("{"), match.numberOfRanges > 1 {
+            return [match.range(at: 1)]
+        } else {
+            return [match.range]
+        }
     }
 
     private func range(for dictionary: SourceKittenDictionary, file: SwiftLintFile) -> NSRange? {

